### PR TITLE
require non-empty inputs to grad() calls in the API

### LINKED ISSUE #47061: When input is empty but not None, `torch.autograd.grad` populates .grad of leaf Tensors
## 🐛 Bug

When input is empty but not None, `torch.autograd.grad` populates .grad of leaf Tensors 

## To Reproduce

The code below fails:
```
x = torch.tensor([1.0], requires_grad=True);
grad = torch.autograd.grad(2*x, [], grad_outputs=torch.tensor([1.0]))
assert(x.grad is None)
```

## Expected behavior

Gradients should not accumulate into leafs for `torch.autograd.grad`

## Environment

PyTorch version: 1.8.0a0+2518abb
Is debug build: True
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: CentOS Linux 8 (Core) (x86_64)
GCC version: (GCC) 8.3.1 20191121 (Red Hat 8.3.1-5)
Clang version: Could not collect
CMake version: version 3.14.0

Python version: 3.8 (64-bit runtime)
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.1
[pip3] torch==1.8.0a0
[conda] blas                      1.0                         mkl
[conda] magma-cuda110             2.5.2                         1    pytorch
[conda] mkl                       2020.2                      256
[conda] mkl-include               2020.2                      256
[conda] mkl-service               2.3.0            py38he904b0f_0
[conda] mkl_fft                   1.1.0            py38h23d657b_0
[conda] mkl_random                1.1.1            py38h0573a6f_0
[conda] numpy                     1.19.1           py38hbc911f0_0
[conda] numpy-base                1.19.1           py38hfa32c7d_0
[conda] torch                     1.8.0a0                   dev_0    <develop>

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved

### PR DESCRIPTION
**Pull request by dagitses**
Summary:
The grad() function needs to return the updated values, and hence
needs a non-empty inputs to populate.

Test Plan:
Passes Python and C++ unit tests, and added new tests to catch this behavior.

Reviewers: alband

Subscribers: soulitzer

Tasks: https://github.com/pytorch/pytorch/issues/47061

Tags:

Fixes #47061

cc @ezyang @gchanan

### CODE REVIEW

File: test/test_jit.py [hunk 1]
```diff
@@ -16092,6 +16092,14 @@ def create_nn_module(*args, **kwargs):
         args_variable, kwargs_variable = create_input(input, dtype=dtype)
         f_args_variable = deepcopy(unpack_variables(args_variable))
 
+        # TODO(issue####) Neither this nor no_grad should be required
+        # if check_against_reference() is updated to check gradients
+        # w.r.t. weights and then only check w.r.t. inputs if any
+        # inputs require it.
+        nonlocal no_grad
```
Comment (albanD) [inline test/test_jit.py:16099]: why do you need the nonlocal here? You want to update this only for the current function right?

Comment (dagitses) [reply test/test_jit.py]: Does it make a difference? The function is only created once per top-level call. Either way, it's sloppy and potentially risky. Does this look better?

Comment (albanD) [reply test/test_jit.py]: Yes much better thanks!

**Review by albanD — APPROVED** [review]
LGTM!

Thanks for fixing the gradcheck!
